lib/ddtrace/contrib/rails/core_extensions.rb

```diff
@@ -1,50 +1,130 @@
+            @tracing_context[:layout] = layout
+          rescue StandardError => e
+            Datadog::Tracer.log.error(e.message)
           end
+
+          # execute the original function anyway
+          render_template_without_datadog(*args)
+        end
+
+        # method aliasing to patch the class
+        alias_method :render_without_datadog, :render
+        alias_method :render, :render_with_datadog
+
+        if klass.private_method_defined?(:render_template) || klass.method_defined?(:render_template)
+          alias_method :render_template_without_datadog, :render_template
+          alias_method :render_template, :render_template_with_datadog
+        else
+          # NOTE: Rails < 3.1 compatibility: the method name is different
+          alias_method :render_template_without_datadog, :_render_template
+          alias_method :_render_template, :render_template_with_datadog
         end
       end
     end
 
-    def patch_renderer_render_partial
-      if defined?(::ActionView::PartialRenderer)
-        ::ActionView::PartialRenderer.class_eval do
-          alias_method :render_partial_without_datadog, :render_partial
-          def render_partial(*args, &block)
-            ActiveSupport::Notifications.instrument('start_render_partial.action_view')
-            render_partial_without_datadog(*args, &block)
-          end
+    def patch_partial_renderer(klass)
+      klass.class_eval do
+        def render_with_datadog(*args, &block)
+          # create a tracing context and start the rendering span
+          @tracing_context = {}
+          ::ActiveSupport::Notifications.instrument(
+            '!datadog.start_render_partial.action_view',
+            tracing_context: @tracing_context
+          )
+          render_without_datadog(*args)
+        rescue Exception => e
+          # attach the exception to the tracing context if any
+          @tracing_context[:exception] = e
+          raise e
+        ensure
+          # ensure that the template `Span` is finished even during exceptions
+          ::ActiveSupport::Notifications.instrument(
+            '!datadog.finish_render_partial.action_view',
+            tracing_context: @tracing_context
+          )
         end
-      else # Rails < 3.1
-        ::ActionView::Partials::PartialRenderer.class_eval do
-          alias_method :render_partial_without_datadog, :render
-          def render(*args, &block)
-            ActiveSupport::Notifications.instrument('start_render_partial.action_view')
-            render_partial_without_datadog(*args, &block)
+
+        def render_partial_with_datadog(*args)
+          begin
+            # update the tracing context with computed values before the rendering
+            template_name = Datadog::Contrib::Rails::Utils.normalize_template_name(@template.try('identifier'))
+            @tracing_context[:template_name] = template_name
+          rescue StandardError => e
+            Datadog::Tracer.log.error(e.message)
```

Comment:
In cases like this – something wrong with the instrumentation itself, not the user's code – I wonder if we shouldn't we use `debug` level.

Reply:
Switching back to `debug` since I really prefer to not spam users logs for any case. Activating the debug mode, is a reasonable way to find out such kind of problems.